File: MPI/shell_sort/shell_sort_parallel.c
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"

#define N 500000
#define ROOT 0
#define TAG 100 /* for MPI_Send */

void rand_init_array(int array[], int n, int upper, int lower);
void display_array(int array[], int n);
void calculate_displacements(int* displacements, int* sendcounts, int size);
void calculate_sendcounts(int* sendcounts, int size, int n);
void shell_sort(int a[], int n);
int* merge(int *x, int n1, int *y, int n2);
int calculate_count(int rank, int* sendcounts, int received);

int main (int argc, char *argv[]) {
    
    int array[N];
    int rank, size;
    MPI_Status status;
    double begin, end;

    MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &size);
	MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    int *sendcounts = malloc(sizeof(int)*size);
    int *displs = malloc(sizeof(int)*size);

    /*  Calculate send counts and displacements (for MPI_Scatterv)  */
    calculate_sendcounts(sendcounts, size, N);
    calculate_displacements(displs, sendcounts, size);

    if (rank == ROOT){
        rand_init_array(array, N, 0, N*4);
        // (void) printf("Initial array: ");
        // display_array(array, N);
        begin = MPI_Wtime();
    }
    
    /* Each thread will have different local_locations vector */
    int* local_array = (int*)malloc(sendcounts[rank]*sizeof(int));
    if (local_array == NULL) {printf ("Memory error\n"); MPI_Finalize(); return 4;}

    MPI_Scatterv(array, sendcounts, displs, MPI_INT, local_array, sendcounts[rank], MPI_INT, ROOT, MPI_COMM_WORLD);

    /* Perform Shell Sort only on the local array with size chunk (every process) */
    shell_sort(local_array, sendcounts[rank]);  

    if (rank == ROOT)
        /* Root process will only send to the next its array */
        MPI_Send(local_array, sendcounts[rank], MPI_INT, rank+1, TAG, MPI_COMM_WORLD);  

    else {  /* if process is not the root */
  
        int received_count = calculate_count(rank, sendcounts, 1);
        int* recv_array = (int*)malloc(received_count*sizeof(int));
        if (recv_array == NULL) {printf ("Memory error\n"); MPI_Finalize(); return 4;}
        
        /* Receive array from previous process */
        MPI_Recv(recv_array, received_count, MPI_INT, rank-1, 100, MPI_COMM_WORLD, &status);

        /* merged_array is the array the process is going to send (size of merged_array: send_count) */
        int send_count = calculate_count(rank, sendcounts, 0);

        /* Merge your array with the received one (merged_array)*/
        int* merged_array = merge(local_array, sendcounts[rank], recv_array, received_count);

        /* Sort merged array */
        shell_sort(merged_array, send_count);

        /* Send the merged array to the next process (when last is last process turn, will send array to root because of modulo) */
        MPI_Send(merged_array, send_count, MPI_INT, (rank+1)%size, TAG, MPI_COMM_WORLD);

    }

    /* Receive & print results */
    if(rank == ROOT) {
        int* results = (int*)malloc(sizeof(int)*N);
        if (results == NULL) {printf ("Memory error\n"); MPI_Finalize(); return 4;}

        MPI_Recv(results, N, MPI_INT, size-1, TAG, MPI_COMM_WORLD, &status);
        end = MPI_Wtime();
        // (void) printf("Sorted Array: ");
        // display_array(results, N);
        (void) printf("Time elapsed: %g\n", (end-begin));
    }

    MPI_Finalize();
    
    return 0;
}

/*
 * Function:  calculate_displacements 
 * --------------------
 * Calculates displacements (for MPI_Scatterv) and puts them on an array
 *
 *  displacements: pointer of the array that will be filled with potential displacements
 *  sendcounts: pointer of the array which contains the size of all local arrays of each process (size of the array is the number of processes)
 *  size: number of processes
 *
 */

void calculate_displacements(int* displacements, int* sendcounts, int size){
    int i, sum=0;
    for (i=0; i < size; ++i) {
        displacements[i] = sum;
        sum += sendcounts[i];
    }
}

/*
 * Function:  calculate_sendcounts 
 * --------------------
 * Calculates sendcounts (how many elements we're going to send to each process, for MPI_Scatterv) and puts them on an array
 *
 *  sendcounts: pointer of the array which contains the size of all local arrays of each process (size of the array is the number of processes)
 *  size: number of processes
 *  n: number of elements of initial array
 *
 */

void calculate_sendcounts(int* sendcounts, int size, int n){
    int i, extra = N % size;
    for (i=0; i < size; ++i) {
        sendcounts[i] = N/size;
        if (extra > 0) {
            ++sendcounts[i];
            --extra;
        }
    }
}

/*
 * Function:  calculate_count 
 * --------------------
 * Calculates send or receive count (how many elements we're going to receive from previous process or send to the next one)
 *
 *  rank: the rank of the process
 *  sendcounts: pointer of the array which contains the size of all local arrays of each process (size of the array is the number of processes)
 *  received: 1 if we are calculating received count, 0 if we are calculating send count
 * 
 * Returns: the asked count (int)
 *
 */

int calculate_count(int rank, int* sendcounts, int received){
    int i, count=0;
    for(i=0;i<=rank-received;++i)
        count += sendcounts[i];

    return count;
}

/*
 * Function:  merge 
 * --------------------
 * Merges two arrays with different length
 *
 *  x: pointer of the first array
 *  n1: size of array x
 *  y: pointer of the second array
 *  n2: size of array y
 * 
 * Returns: pointer of the merged array
 *
 */

int* merge(int *x, int n1, int *y, int n2){
	int i=0, j=0, k=0;
	int* result = (int*)malloc((n1+n2)*sizeof(int));;

	while(i < n1 && j < n2)
		if (x[i] < y[j]){
			result[k] = x[i];
			++i; 
            ++k;
		}
		else{
			result[k] = y[j];
			++j; 
            ++k;
		}
	if (i == n1)
		while(j < n2){
			result[k] = y[j];
			++j;
            ++k;
		}
	else
		while(i < n1){
			result[k] = x[i];
			++i;
            ++k;
		}

	return result;
}

/*
 * Function:  rand_init_array 
 * --------------------
 * Fills an integer array with random numbers
 *
 *  array: the array that will be filled with numbers
 *  n: number of elements in the array
 *  upper: highest value of random number
 *  lower: lowest value of random number
 *
 */

void rand_init_array(int array[], int n, int upper, int lower){
    int i;    
    for (i=0; i<n; ++i)
        array[i] = (rand() % (upper - lower + 1)) + lower;
}

/*
 * Function:  display_array 
 * --------------------
 * Prints an integer array to user
 *
 *  array: the array that will be printed
 *  n: number of elements in the array
 *
 */

void display_array(int array[], int n){
    (void) printf("[ ");
    int i;
    for (i=0; i<n; ++i)
        (void) printf("%d ", array[i]);
    (void) printf("]\n\n");
}

/*
 * Function:  shell_sort 
 * --------------------
 * Sorts an integer array, using the shell sort algorithm
 *
 *  a: the array that will be sorted
 *  n: number of elements in the array
 *
 */

void shell_sort(int a[], int n){
    int h, i, j, t;
    for (h=n; h /= 2;) {
        for (i=h; i < n; ++i) {
            t = a[i];
            for (j=i; j >= h && t < a[j - h]; j -= h)
                a[j] = a[j - h];
            a[j] = t;
        }
    }
}

File: MPI/mpi4py/shell_sort/shell_sort_parallel.py

import numpy as np
from mpi4py import MPI

TAG = 200
ROOT = 0
N = 10
LOWER = 0
UPPER = N*2

def calculate_displacements(sendcounts):
    '''
    Function:  calculate_displacements 
    --------------------
    Calculates displacements (for MPI_Scatterv) and puts them on an array
    
    sendcounts: pointer of the array which contains the size of all local arrays of each process (size of the array is the number of processes)
    '''
    return np.insert(np.cumsum(sendcounts), 0, 0)[0:-1]

def calculate_sendcounts(split):
    '''
    Function:  calculate_sendcounts 
    --------------------
    Calculates sendcounts (how many elements we're going to send to each process, for MPI_Scatterv) and puts them on an array
 
    split: pointer of the array which contains the split of initial array (split is done using numpy's split)
    '''
    return [len(split[i]) for i in range(len(split))]

def merge(x, n1, y, n2):
    '''
    Function:  merge 
    --------------------
    Merges two arrays with different length

    x: pointer of the first array
    n1: size of array x
    y: pointer of the second array
    n2: size of array y
    
    Returns: pointer of the merged array
    '''
    i=0
    j=0
    k=0
    result = np.empty(dtype=int, shape=n1+n2)
    while i < n1 and j < n2:
        if x[i] < y[j]:
            result[k] = x[i]
            i += 1
            k += 1
        else:
            result[k] = y[j]
            j += 1
            k += 1
    if i == n1:
        while j < n2:
            result[k] = y[j]
            j += 1
            k += 1
    else:
        while i < n1:
            result[k] = x[i]
            i += 1
            k += 1

    return result

def shell_sort(array, n):
    '''
    Function:  shell_sort 
    --------------------
    Sorts an integer array, using the shell sort algorithm

    a: the array that will be sorted
    n: number of elements in the array
    '''
    h = int(n/2)
    while h > 0:
        for i in range (h, n):
            t = array[i]
            j=i
            while j >= h and t < array[j-h]:
                array[j] = array[j - h]
                j -= h
            array[j] = t
        h = int(h/2)
    
def calculate_count(sendcounts, rank):
    sum = 0
    for i in range(rank+1):
        sum += sendcounts[i]
    return sum

if __name__ == '__main__':

    comm = MPI.COMM_WORLD
    size = comm.Get_size()
    rank = comm.Get_rank()
    status = MPI.Status()

    array = np.empty(N, np.int)
    if rank == ROOT:
        array = np.random.randint(low=LOWER, high=UPPER, size=N)
        # root makes the split
        split = np.array_split(array, size)
        print(f'\nInitial Array: {array}')
        begin = MPI.Wtime()
        my_chunk = split[rank]
        # Every process has split[rank] chunk :-))
        for i in range(1,size):
            comm.send(list(split[i]), i, TAG)   # Root sends to each process
    else:
        my_chunk = comm.recv(source=ROOT, tag=TAG, status=status)   # The others receive

    # Each thread will have different local_locations vector 
    local_array = np.asarray(my_chunk)

    # Perform Shell Sort only on the local array with size chunk (every process)
    shell_sort(local_array, local_array.size)

    # Each thread will have different local_locations vector 
    if rank == ROOT:
        # Root process will only send to the next its array
        comm.send(obj=local_array.size, dest=rank+1, tag=TAG)
        comm.Send(buf=local_array, dest=rank+1, tag=TAG)

    else:  # if process is not the root

        # Receive the size of rank-1 array
        recv_count = comm.recv(source=rank-1, tag=TAG, status=status)
        
        # Make a recv array object
        recv_array = np.empty(dtype=np.int, shape=recv_count)

        # Receive array from previous process
        comm.Recv(buf=recv_array, source=rank-1, tag=TAG, status=status)

        # Merge your array with the received one (merged_array)
        merged_array = merge(local_array, local_array.size, recv_array, recv_count)

        comm.send(obj=merged_array.size, dest=(rank+1)%size, tag=TAG)
        # Send the merged array to the next process (when last is last process turn, will send array to root because of modulo)
        comm.Send(buf=merged_array, dest=(rank+1)%size, tag=TAG)

    # Receive & print results 
    if rank == ROOT:
        # Root receives merged_array from last process (size == N)
        recv_count = comm.recv(source=size-1, tag=TAG, status=status)
        # Results vector
        results = np.empty(shape=recv_count, dtype=np.int)
        # Receive final vector
        comm.Recv(buf=results, source=size-1, tag=TAG, status=status)
        end = MPI.Wtime()
        print(f'\nSorted Array: {results}')
        print(f'\nTime elapsed for operation: {end-begin} seconds.\n')